### PR TITLE
feat(helm): add --app-version flag to 'helm install/upgrade'

### LINKED ISSUE #3555: Add 'helm install --app-version' command/versioning a chart against an app version
As far as I'm aware, `helm install` currently only supports specifying the `--version` flag to specify a chart version to install.

I'm unsure how the 'appVersion' field in a Chart.yaml is supposed to be used, but it seems generally useful to add support for versioning your application against a specific version (or version set) of a Chart.

Am I mis-using the `appVersion` field here? Should I instead be constantly building my chart to be backward compatible with previous version, or otherwise how can I infer to my users which chart version to specify when running `helm install` if they want a particular version (this becomes even more complex when you consider a user can also change the version deployed with something like `--set image.tag`, which often results in a version change of an application).

### PR DESCRIPTION
**Pull request by eraac**
Closes #3555

New pull request for https://github.com/helm/helm/pull/4961, sorry about that

**If applicable**:
- [X] this PR contains documentation
- [x] this PR contains unit tests

### CODE REVIEW

**Comment by Eraac** [comment]
That commit include some change according to https://github.com/helm/helm/pull/4961#issuecomment-466759140 and adding new boolean into Chart metadata (by default false) for allowing or not the flag --app-version

**Comment by marshallford** [comment]
Is this gtg?

**Comment by Eraac** [comment]
> Is this gtg?

@marshallford yes, waiting for a review at least

File: cmd/helm/upgrade.go [hunk 1]
```diff
@@ -247,23 +249,32 @@ func (u *upgradeCmd) run() error {
...
 		}
-	} else {
-		return prettyError(err)
+	} else if err != chartutil.ErrRequirementsNotFound {
+		return fmt.Errorf("cannot load requirements: %v", err)
+	}
+
+	if u.appVersion != "" {
+		if !chart.Metadata.OverrideAppVersion {
+			return fmt.Errorf("override app version is not allowed on this chart")
```
Comment (bacongobbler) [inline cmd/helm/upgrade.go]: Can you please improve the error message to say *what* the chart developer has to do to allow the app version to be overridden? We need a call-to-action so the user knows what they need to do to override the app version.

**Review by bacongobbler — CHANGES_REQUESTED** [review]
Thank you for the PR.

Before we can consider reviewing this further, would you please document how chart developers can allow users to set the `--app-version` flag for their chart. From looking at this code, by default, any chart that omits the `overrideAppVersion` field (i.e. all charts of the legacy charts) will be unable to use this unless they opt in. Users need to know what they need to do to allow users to override the app version field in their charts.

If you're thinking of porting this over to Helm 3, perhaps @adamreese should take a look at this PR. I'm still concerned about the overall design of allowing users to override chart metadata as part of `helm install`.

Thanks!

File: cmd/helm/upgrade.go (hunk 1 above)
Comment (bacongobbler) [reply cmd/helm/upgrade.go]: I'd suggest something like:

> The chart maintainer has disallowed overriding the chart's package metadata. Please contact the maintainer to set `overrideChartMeta` in their Chart.yaml to `true`.

File: _proto/hapi/chart/metadata.proto [hunk 2]
```diff
@@ -90,4 +90,7 @@ message Metadata {
 
 	// KubeVersion is a SemVer constraint specifying the version of Kubernetes required.
         string kubeVersion = 17;
+
+	// Whether or not the app version can be override by flag --app-version (during install or update)
```
Comment (bacongobbler) [inline _proto/hapi/chart/metadata.proto]: please change this to:

> // During chart install or update, this field determines whether or not the metadata can be overridden by flags like --app-version

File: _proto/hapi/chart/metadata.proto [hunk 3]
```diff
@@ -90,4 +90,7 @@ message Metadata {
...
+	bool overrideAppVersion = 18;
```
Comment (bacongobbler) [inline _proto/hapi/chart/metadata.proto]: please change this to `overrideMetadata`; there may be more fields in the future that the user may wish to override, and I don't see a reason why we need to create `override` fields for each field we want to allow the user to override.

**Review by bacongobbler — CHANGES_REQUESTED** [review]

**Comment by Eraac** [comment]
> If you're thinking of porting this over to Helm 3, perhaps @adamreese should take a look at this PR. I'm still concerned about the overall design of allowing users to override chart metadata as part of helm install.

@bacongobbler I haven't look next major version yet, what I should know about it ? Of course, i think porting this in helm 3

By the way I have make changes according to the review

**Comment by dmydlarz** [comment]
Hello, what is the status of this PR?

**Comment by Stono** [comment]
would still _really_ like this feature.

**Comment by unitto1** [comment]
Any updates about merging this? This is a very useful feature, hope to get it =)

**Comment by Eraac** [comment]
Circleci fail, but i have no idea why. If someone has explication, i can fix it (i guess)

**Comment by bacongobbler** [comment]
```
#!/bin/bash -eo pipefail
.circleci/test.sh
Running 'make build'
GOBIN=/go/src/k8s.io/helm/bin go install  -tags '' -ldflags '-w -s -X k8s.io/helm/pkg/version.GitCommit=b5f34b33bb0a61800ed83066f9654c5b5fa29ed0 -X k8s.io/helm/pkg/version.GitTreeState=clean' k8s.io/helm/cmd/...
Running 'make verify-docs'
GOBIN=/go/src/k8s.io/helm/bin go install  -tags '' -ldflags '-w -s -X k8s.io/helm/pkg/version.GitCommit=b5f34b33bb0a61800ed83066f9654c5b5fa29ed0 -X k8s.io/helm/pkg/version.GitTreeState=clean' k8s.io/helm/cmd/...
Creating /root/.helm 
Creating /root/.helm/repository 
Creating /root/.helm/repository/cache 
Creating /root/.helm/repository/local 
Creating /root/.helm/plugins 
Creating /root/.helm/starters 
Creating /root/.helm/cache/archive 
Creating /root/.helm/repository/repositories.yaml 
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com 
Adding local repo with URL: http://127.0.0.1:8879/charts 
$HELM_HOME has been configured at /root/.helm.
Not installing Tiller due to 'client-only' flag having been set
--- /tmp/tmp.xDUewHmInp/docs/helm/helm_install.md	2019-06-07 13:13:04.965128551 +0000
+++ docs/helm/helm_install.md	2019-06-07 13:10:32.990535551 +0000
@@ -83,7 +83,7 @@ helm install [CHART] [flags]
       --ca-file string           Verify certificates of HTTPS-enabled servers using this CA bundle
       --cert-file string         Identify HTTPS client using this SSL certificate file
       --dep-up                   Run helm dependency update before installing the chart
-      --description string       Specify a description for the release
+      --description string       specify a description for the release
       --devel                    Use development versions, too. Equivalent to version '>0.0.0-0'. If --version is set, this is ignored.
       --dry-run                  Simulate an install
   -h, --help                     help for install
helm docs are out of date. Please run "make docs"
Makefile:150: recipe for target 'verify-docs' failed
make: *** [verify-docs] Error 1
Exited with code 2
```

**Comment by Eraac** [comment]
Thanks @bacongobbler, CircleCI pass successfully now, and no conflict 🎉  
What's the next step?

**Comment by bacongobbler** [comment]
I'd suggest reaching out and finding another maintainer to review this. I'm uncomfortable merging a PR proposing a design that I don't believe is the right path forward for the chart spec as per my last comment made in https://github.com/helm/helm/pull/4961#issuecomment-456152753.

**Review by Starefossen — APPROVED** [review]

**Comment by iorlas** [comment]
Okay. I've been upgrading to the helm this week, since tiller was finally ditched and 3.0alpha is released.

And I have encountered the same problems as all you, guys. And I faced the fact that I can't use appVersion in my workflow. I need ability to set appVersion when I `helm install` it. But on the second thought... it is so wrong, I cannot just ignore this pull request. Let me elaborate. 

### Helm install

First, we need to understand where the problem starts.

Actually, helm install has 2 behaviours:

1. Install remote package

Feels nice and play, similar to the apt, pip, gem, yum, npm - any package manager. Even newbie will handle helm here.

2. Install local ... wait a minute!

When we install the local chart, it is not a package yet. As it is not a package and we may not need to make one, do we need to edit anything? Maybe... we need to install it directly, every time, since it is handy. This is where the problem begins.

### The Use Case

User has a project. User has CI/CD. CD manages the versioning process. User does not want the repo thing, since project is private or personal. User wants to `helm install` it.

User wants to use `Chart.appVersion`, but he is unable to edit it within CD flow, but user won't edit it manually, since it feels redundant.

User is stuck.

### Options

User has options:

1. Ignore `Chart.appVersion` and use `Values.image.tag`/`Values.appVersion`

@bacongobbler gave us a good snippet:
```
{{ default .Values.appVersion .Chart.AppVersion }}
```
Common approach, but feels weird, since we will have two appVersions, it is not conventional, makes harder to understand the codebase for a newbie. It creates a variety lf approaches for the same configurational problem, which should be simple and convenient, with one source of appVersion

2. Force the helm team to add ability to edit `Chart.yaml` using `helm install` arguments

It solves the problem in much more convenient approach when the first one: you will use the same variable, same flow, everything is perfect and shiny! **But**, it will mangle whole `helm install` thing with package-related arguments, will lead to possible bugs with override mechanics and, more important, break the logic and open a path for proposals to add more wild arguments to the install, leading to the hell of the bug reports and increased learning curve.

### Immutability

I totally agree with the points @bacongobbler gave. First, immutability. If we will add arguments like so, what is the difference would be between the values and chart yaml files? What logic would be taken when someone will propose to add one more argument, then one more from the values? It'll be a chaos.

If we need `Chart.yaml` and `values.yaml`, we need to know why it is split. And we have some reasons:

Chart.yaml - for release variables, which are bound to the release as artifacts, like appVersion, which is used to glue multiple remote artifacts together. So whole ChartPackage will be a release, like a black box. It should not be accessible when we "install the software"

Values.yaml - variables and its defaults. Like arguments for the black box, for a function. It could be supplied, could be overridden. I would even call it `arguments.yaml` or `defaults.yaml`

Second, we already have approach which solves the problem, @bacongobbler snippet:

```
{{ default .Values.appVersion .Chart.AppVersion }}
```

### The real problem

The real problem is, `helm install` is used wrong. More than that, it is implemented wrong. It should not allow to install the Chart Template, it should ask for a package instead.

Using `helm install` as `helm package && helm install` leads to the proposals to add `helm package` arguments to the `helm install`. **And we already have `--app-version` argument in the `helm package` command**! Why on earth you would need it anywhere else?!

### Proposal

1. `helm install` should work only with packages
2. `helm package-and-install` **could** be implemented as a shortcut, but it will need to accept both package and install arguments

**Comment by Eraac** [comment]
`helm package` doesn't allow to move the helm chart outside the repository (which is bad because developer can update all the helm-chart)

**Comment by wasker** [comment]
I'd like to add another vote for incorporating this change. We rarely change the chart itself, and only vary image tag and a bunch of meta-information about deployment. In our scenario `helm list` is not helpful by showing the version that baked into the chart vs. what's actually there.

Please accept this change.

**Comment by bacongobbler** [comment]
@iorlas I like your breakdown and your suggestions. Thank you for your insight and feedback, and you make a perfect point RE: this case has already been solved if you `helm package --app-version 0.1.0 ./mychart && helm install ./mychart --generate-name`. Perhaps we should consider that as the path forward.

**Comment by iorlas** [comment]
> `helm package` doesn't allow to move the helm chart outside the repository (which is bad because developer can update all the helm-chart)

**Problem**: helm package packages files into archive, so if we will package it and shove into the repository, it won't be easy to customise it, since you will need to extract it first, then package again and then publish.

@Eraac, am I correct in understanding problem you describe?
What is the repository you are talking about? Helm repo or git repo? Since Helm repo is deprecated right now in favor of k8s facility, I assume you are talking about git repos.

So, again, we are not talking about the real problem. Currently Helm consists of multiple idioms. Especially: Chart, Package.

## Package

Package is a build artifact, release, application. Package is supposed to be customisable using variables(set/values).

Treat it like a build: you have a binary or minimised js/css.

Think for a minute, **are your application builds supposed to be customisable** more than configuration allows? No

## Chart

Chart is a template. It is like a source code. You can customise it, take some parts out of it, add another.

But when you change the source code, are you complaining you are not able to run is directly on you production environment, w/o compiling it first? _We are able in some stacks, but under the hood it is the same compile-then-run, so you won't be able to run it w/o compiling_

## Sooo....

The real problem all you guys are struggling: you don't want to use Package idiom. You want to skip it when you are working on the non-public project, so it doesn't make sense to "create some reusable package"

**And I'm with you all on it!**

_Although I feel like it is solid, clean way to distinguish build-specific and stage-specific configuration. I create package as build artifact, then installing it on multiple stages._

**Use-case**: You have small, pet-project, even a startup, you don't want to deal with such complications

**Problem**: You see **Helm as orchestration utility**, which consists of: **templates**, **input values**. And for you, there are no reasons to have anything beside that!

## Reasoning

As per my previous comment, `helm package && helm install` is a way Helm works. **There are no reasons to change that**. It will simplify Helm a lot, but will leave packaging mechanism w/o many important features, leading to the more complicated process of installing foreign packages.

And, this exact pull request will break distinguish between package and install. It will make harder for newbies to learn the Helm. **It will break the Helm!**

## Proposal

@bacongobbler, we need to simplify Helm, we need to support both medium/big and small projects. Small project don't need that. Project app could run the code using `go run` in production and it is fine, since development time does matter to test the idea.

In this effort, I would suggest to add **new command**

## `helm run`

**Arguments**: `helm package` + `helm install`, except some non-viable for such usage, like --repo or --devel, up to discussion

**Function**: render the templates state using all arguments, install it directly w/o creating the package archive

It will behave like `helm install`, simplifying it in exchange. It will work as `helm package`, but giving working k8s resources release as a product.

**Comment by iorlas** [comment]
By the way, there is one use-case described here https://github.com/helm/helm/issues/3555#issuecomment-511730672

> I ran into something where I need to pin to version 0.6.0 of an app, and the chart version has no relation to the app version.

But I see that `helm run` or `helm package && helm install` should solves it.

**Comment by wasker** [comment]
In our case it's not a "pet project", but a bunch of microservices that are following the same template (code, CI/CD, k8s infra) and share a bunch of CI/CD and k8s setup bits. All of them are owned by different teams and are deployed on separate schedules in same or different k8s clusters. There's really no need in publishing Helm packages, because those packages won't be consumed outside of CI/CD context which builds each microservice. Variables that are important for our scenario are: chart version (i.e. version of the template that is used to install the microservice) and an app version (i.e. version of the microservice which was installed with the template).

**Comment by iorlas** [comment]
@wasker Is isn't pet project, but it still needs "simplified" behavior: write some scripts, install/upgrade it directly. Am I right? It is what I'm trying to describe.

Still, package is not necessary sharable artifact, but it is the way to deliver your app into your cluster. Publishing is a like an additional feature allowed by such design.

How you set chart version? Are you ok with setting it manually? It supposed to be, since it is the "scrips version".

What would you think if you had such command as I described, `helm run`? So, `helm install` would install some remote/local package from the repo, but `helm run` will install your scripts w/o packaging, allowing to set all variables as cmd parameters?

**Comment by wasker** [comment]
@iorlas I'm not attached to the actual command we'd have to use for that.

**Comment by gabemeola** [comment]
A "package and install" command would be very useful and simplify the workflow a bit. Currently I need to find the name and version of the packaged file. Having it "in one go" would be nice.

**Comment by iorlas** [comment]
>Currently I need to find the name and version of the packaged file. Having it "in one go" would be nice.

Same thing here, so we had to hardcode in CI variables, changing it manually, otherwise it won't deploy.

I'm not helm contributor/maintainer, but maybe I'll be able to dedicate some time later this month for PoC. We really need it...

**Comment by BenjaminSchiborr** [comment]
Here are my 2 cents on this:
Assuming we have a helm chart version X deploying a service appVersion Y. 

The helm chart is used to describe the infrastructure inside of Kubernetes on version X which is used to host a service on appVersion Y inside of it. During initial development both X and Y will change regularly. However, at some point X will be more or less stable and Y will continue to change (unless Y has some new requirements to infrastructure, which most likely happens far less often in the development cycle of Y).

With the original approach proposed in this ticket one could take a stable helm chart version X to deploy an appVersion Y, Y+1, Y+N, etc. 

However, not allowing this flag to be overridden during helm install and instead only in package would effectively tie both X and Y together forcing me to always creating a new X+1 for Y+1. This seems unnecessary to me and would result in a ton of helm packages that effectively haven't changed apart from that they are referencing a new appVersion.

**Comment by bacongobbler** [comment]
Closing as we're not planning to move this PR forward. A command that wraps the logic of `helm package` and `helm install` would solve the use case originally described. Let's carry on with that discussion back in #3555. Thanks everyone!

**Comment by FloLaco** [comment]
According to @iorlas and @bacongobbler , there is a workaround if we use `{{ default .Values.appVersion .Chart.AppVersion }}`. Maybe it's a stupid question, but where to use it ? `Chart.yaml` does not accept any function.

`Error: cannot load Chart.yaml: error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{"default .Values.appVersion .Chart.AppVersion":interface {}(nil)}`

**Pull request closed without merging**